assignment2/commonAncestor.py

```diff
@@ -0,0 +1,48 @@
+"""Design an algorithm and write code to find the lowest common ancestor of two
+nodes in a binary tree. Avoid storing additional nodes in a data structure."""
+
+from binarytree import BinaryTree
+
+def commonAncestor(tree, key1, key2):
+    """Find the lowest common ancestors of the given nodes in the given Binary
+    Tree.
+
+    Args:
+        tree: a BinaryTree;
+        key1: an integer, key of the first node;
+        key2: an integer, key of the second node.
+
+    Return:
+        an integer, key of the lowest common ancestors of two given nodes,
+        or `None` if some of the given nodes doesn't exists in the given tree.
+    """
+    def _commonAncestor(cur):
```

Comment:
This implementation is rather convoluted: the way the logic of whether 0, 1, or 2 of the keys were found in each subtree is not straightforward. I think it might be helpful to make the code more explicit (even if maybe less efficient).

My main point is that since it is not straightforward to understand how it works, if yourself or someone else needs to modify this code at a later time, it is easy to make a mistake.

Maybe start with something simpler:

if cur.key == key1 == key2 you can return right away
if cur.key == key1 and you can find key2 under cur.left or cur.right, you can return as well.
and enumerate simpler cases one by one.